### ISSUE TITLE
jsonpickle.decode exhibits different behavior in AREPL than in Pycharm

### ISSUE THREAD

**Issue by Almenon** [Almenon/AREPL#10]
In AREPL json.decode fails to decode a simple pickled enum.  In Pycharm the decoding happens without any error.  The reason behind this is that AREPL uses exec to execute the user's code.  In the example below you can see the exact same code that works outside of exec() fails inside exec().

```python
from enum import Enum
from jsonpickle import encode, decode


class C(Enum):
    VALUE = 1

my_list_encode = encode(C.VALUE)
decode(my_list_encode) # no error

execCode = """ # exact same code as above follows!
from enum import Enum
from jsonpickle import encode, decode


class C(Enum):
    VALUE = 1

my_list_encode = encode(C.VALUE)
decode(my_list_encode)
"""

execLocals = {}
exec(execCode,execLocals) # error!
```

> Traceback (most recent call last):
  File "C:/dev/random python scripts/misc.py", line 24, in <module>
    exec(execCode,execLocals)
  File "<string>", line 10, in <module>
  File "C:\dev\AREPL\src\python\jsonpickle\__init__.py", line 152, in decode
    return unpickler.decode(string, backend=backend, keys=keys)
  File "C:\dev\AREPL\src\python\jsonpickle\unpickler.py", line 27, in decode
    return context.restore(backend.decode(string), reset=reset)
  File "C:\dev\AREPL\src\python\jsonpickle\unpickler.py", line 120, in restore
    value = self.\_restore(obj)
  File "C:\dev\AREPL\src\python\jsonpickle\unpickler.py", line 162, in \_restore
    return restore(obj)
  File "C:\dev\AREPL\src\python\jsonpickle\unpickler.py", line 183, in \_restore_reduce
    if f == tags.NEWOBJ or f.\_\_name__ == '\_\_newobj\_\_':
AttributeError: 'dict' object has no attribute '\_\_name\_\_'
"""

**Comment by Almenon** [comment]
the problem is fixed when you replace execLocals = {} with execLocals = locals().  It must be something missing in locals that causes the error.

**Comment by Almenon** [comment]
Found it!  execLocals was missing the \_\_name\_\_ property.  Inserting the following code after execLocals fixes the issue:

```python
execLocals['__doc__'] = locals()['__doc__']
execLocals['__file__'] = locals()['__file__']
execLocals['__loader__'] = locals()['__loader__']
execLocals['__name__'] = locals()['__name__']
execLocals['__package__'] = locals()['__package__']
execLocals['__spec__'] = locals()['__spec__']
```

**Comment by Almenon** [comment]
When I tried applying the fix to the pythonEvaluator the fix did not work.  I reinvestigated and found out that when you remove the block of code in the beginning:

```python
from enum import Enum
from jsonpickle import encode, decode

class C(Enum):
    VALUE = 1

my_list_encode = encode(C.VALUE)
decode(my_list_encode) # no error
```

then the fix does not work anymore!  I have no idea why the lack of this code would cause an error in execd code - they don't share the same locals.  But I have already spent too much time on this - I decided to just remove the v1.0.0-alpha.2 milestone and commit what I had so far.  Adding the special vars (\_\_doc\_\_, etc...) is a good idea that should've been done already, even if it didn't fix this particular problem.

Also, I found out that the code block above does not work in the console.  This means something funky is going on in jsonpickle - possibly if I raise an issue with them the fix would also fix the error in my case.

I'll look into this issue again if I have time to investigate it further or someone comments.  For now I have more important stuff to do.
